src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs

```diff
@@ -0,0 +1,337 @@
+                // Spawn send and receive logic
+                var receiveTask = DoReceive();
+                var sendTask = DoSend();
+
+                // Now wait for both to complete
+                await receiveTask;
+                await sendTask;
+
+            }
+            catch (Exception ex)
+            {
+                _log.LogError(0, ex, $"Unexpected exception in {nameof(QuicStreamContext)}.{nameof(StartAsync)}.");
+            }
+        }
+
+        private async Task DoReceive()
+        {
+            Exception error = null;
+
+            try
+            {
+                await ProcessReceives();
+            }
+            catch (QuicException ex)
+            {
+                // This could be ignored if _shutdownReason is already set.
+                error = new ConnectionResetException(ex.Message, ex);
+            }
+            catch (Exception ex)
+            {
+                // This is unexpected.
+                error = ex;
+                _log.StreamError(ConnectionId, error);
+            }
+            finally
+            {
+                // If Shutdown() has already bee called, assume that was the reason ProcessReceives() exited.
+                Input.Complete(_shutdownReason ?? error);
+
+                FireStreamClosed();
+
+                await _waitForConnectionClosedTcs.Task;
+            }
+        }
+
+        private async Task ProcessReceives()
+        {
+            var input = Input;
+            while (true)
+            {
+                var buffer = Input.GetMemory(MinAllocBufferSize);
+                var bytesReceived = await _stream.ReadAsync(buffer);
+
+                if (bytesReceived == 0)
+                {
+                    // Read completed.
+                    break;
+                }
+
+                input.Advance(bytesReceived);
+
+                var flushTask = input.FlushAsync();
+
+                var paused = !flushTask.IsCompleted;
+
+                if (paused)
+                {
+                    _log.StreamPause(ConnectionId);
+                }
+
+                var result = await flushTask;
+
+                if (paused)
+                {
+                    _log.StreamResume(ConnectionId);
+                }
+
+                if (result.IsCompleted || result.IsCanceled)
+                {
+                    // Pipe consumer is shut down, do we stop writing
+                    break;
+                }
+            }
+        }
+
+        private void FireStreamClosed()
+        {
+            // Guard against scheduling this multiple times
+
+            if (_streamClosed)
+            {
+                return;
+            }
+
+            _streamClosed = true;
+
+            ThreadPool.UnsafeQueueUserWorkItem(state =>
+            {
+                state.CancelConnectionClosedToken();
+
+                state._waitForConnectionClosedTcs.TrySetResult(null);
+            },
+            this,
+            preferLocal: false);
```

Comment:
In this case, I don't think there's a reason not to prefer the local thread

Reply:
Why would this situation vary from the SocketConnection? Should we have preferLocal: true as well there?

Reply:
This situation doesn't vary from SocketConnection. Looks like this found it's way into SocketConnection as part of [a rather large PR](https://github.com/dotnet/aspnetcore/pull/10321) and got missed.